pydarn/io/borealis/borealis_array.py

```diff
@@ -0,0 +1,482 @@
+# Copyright 2019 SuperDARN Canada, University of Saskatchewan
+# Author: Marci Detwiller
+"""
+This file contains classes for reading, and 
+writing of Borealis array file types. Array file 
+types are restructured Borealis files. Files are restructured
+after being written by the radar to condense all information 
+from that time period into arrays. The other type of 
+Borealis file is the site file. Site file types
+are Borealis site files, ie. stored in a record-by-record
+fashion, before being converted to array types.
+
+Classes
+-------
+BorealisArrayRead: Reads array-style Borealis SuperDARN file types (hdf5). 
+    These are the files commonly distributed and available.
+BorealisArrayWrite: Writes array-style Borealis SuperDARN file types (hdf5).
+    These are the files commonly distributed and available.
+
+Exceptions
+----------
+BorealisFileTypeError
+BorealisFieldMissingError
+BorealisExtraFieldError
+BorealisDataFormatTypeError
+BorealisNumberOfRecordsError
+
+See Also
+--------
+BorealisSiteRead
+BorealisSiteWrite
+borealis_site_to_array_file
+borealis_array_to_site_file
+
+For more information on Borealis data files and how they convert to dmap,
+see: https://borealis.readthedocs.io/en/latest/ 
+
+"""
+import deepdish as dd
+import h5py
+import logging
+import math
+import numpy as np
+import os
+
+from collections import OrderedDict
+from datetime import datetime
+from typing import Union, List
+
+from pydarn import borealis_exceptions, borealis_formats, \
+                   BorealisUtilities, borealis_array_to_site_dict
+from pydarn.utils.conversions import dict2dmap
+
+pydarn_log = logging.getLogger('pydarn')
+
+
+class BorealisArrayRead():
+    """
+    Class for reading Borealis 'array' filetypes. Array files have
+    been restructured to no longer be read in record-by-record
+    style. 
+
+    See Also
+    --------
+    BorealisRawacf
+    BorealisBfiq
+    BorealisAntennasIq
+    BorealisRawrf
+
+    Attributes
+    ----------
+    filename: str
+        The filename of the Borealis HDF5 file being read.
+    borealis_filetype: str
+        The type of Borealis file. Restructured types include:
+        'bfiq'
+        'antennas_iq'
+        'rawacf'
```

Comment:
Rawrf has no restructured format at this time. It is rarely produced so I don't think it would be necessary to have restructuring built for it, but maybe in the future.